Sources/InteroperabilityTests/InteroperabilityTestCases.swift

```diff
@@ -0,0 +1,721 @@
+///    ```
+///    {
+///      response_size: 314159
+///      payload:{
+///        body: 271828 bytes of zeros
+///      }
+///    }
+///    ```
+/// 2. The client attaches custom metadata with the following keys and values
+///    to a FullDuplexCall with request:
+///    - key: "x-grpc-test-echo-initial", value: "test_initial_metadata_value"
+///    - key: "x-grpc-test-echo-trailing-bin", value: 0xababab
+///    ```
+///    {
+///      response_parameters:{
+///        size: 314159
+///      }
+///      payload:{
+///        body: 271828 bytes of zeros
+///      }
+///    }
+///    ```
+///    and then half-closes
+///
+/// Client asserts:
+/// - call was successful
+/// - metadata with key "x-grpc-test-echo-initial" and value "test_initial_metadata_value" is
+///   received in the initial metadata for calls in Procedure steps 1 and 2.
+/// - metadata with key "x-grpc-test-echo-trailing-bin" and value 0xababab is received in the
+///   trailing metadata for calls in Procedure steps 1 and 2.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+struct CustomMetadata: InteroperabilityTest {
+  let initialMetadataName = "x-grpc-test-echo-initial"
+  let initialMetadataValue = "test_initial_metadata_value"
+
+  let trailingMetadataName = "x-grpc-test-echo-trailing-bin"
+  let trailingMetadataValue: [UInt8] = [0xAB, 0xAB, 0xAB]
+
+  func checkInitialMetadata(_ metadata: Metadata) throws {
+    try assertEqual(metadata.count, 1)
+    var metadataIterator = metadata.makeIterator()
+    guard let initialMetadataPair = metadataIterator.next() else {
+      try assertFailure("The initial metadata should contain a key value pair.")
+      return
+    }
+    try assertEqual(initialMetadataPair.key, self.initialMetadataName)
+    try assertEqual(initialMetadataPair.value, Metadata.Value.string(self.initialMetadataValue))
+  }
+
+  func checkTrailingMetadata(_ metadata: Metadata) throws {
+    try assertEqual(metadata.count, 1)
+    var metadataIterator = metadata.makeIterator()
+    guard let trailingMetadataPair = metadataIterator.next() else {
+      try assertFailure("The trailing metadata should contain a key value pair.")
+      return
+    }
+    try assertEqual(trailingMetadataPair.key, self.trailingMetadataName)
+    try assertEqual(trailingMetadataPair.value, Metadata.Value.binary(self.trailingMetadataValue))
+  }
+
+  func run(client: GRPCClient) async throws {
+    let testServiceClient = Grpc_Testing_TestService.Client(client: client)
+
+    let unaryRequest = Grpc_Testing_SimpleRequest.with { request in
+      request.responseSize = 314_159
+      request.payload = Grpc_Testing_Payload.with {
+        $0.body = Data(count: 271_828)
+      }
+    }
+    var metadata = Metadata()
+    metadata.addString(self.initialMetadataValue, forKey: self.initialMetadataName)
+    metadata.addBinary(self.trailingMetadataValue, forKey: self.trailingMetadataName)
+    let streamingMetadata = metadata
```

Comment:
You can use the fact that `Metadata` is `ExpressibleByDictionaryLiteral` here to avoid the mutable `metadata`:

```swift
let metadata: Metadata = [
  self.initialMetadataName: .string(self.initialMetadataValue),
  self.trailingMetadataName: .binary(self.trailingMetadataValue)
]
```